src/physics/reynolds_averaged_navier_stokes.cpp

```diff
@@ -0,0 +1,823 @@
+    // Templated virtual member functions
+    dealii::Tensor<2,dim,real2> viscous_stress_tensor;
+    dealii::Tensor<1,dim,real2> heat_flux;
+    if constexpr(std::is_same<real2,real>::value){ 
+        viscous_stress_tensor = compute_Reynolds_stress_tensor(primitive_soln_rans, primitive_soln_gradient_rans,primitive_soln_turbulence_model);
+        heat_flux = compute_Reynolds_heat_flux(primitive_soln_rans, primitive_soln_gradient_rans,primitive_soln_turbulence_model);
+    }
+    else if constexpr(std::is_same<real2,FadType>::value){ 
+        viscous_stress_tensor = compute_Reynolds_stress_tensor_fad(primitive_soln_rans, primitive_soln_gradient_rans,primitive_soln_turbulence_model);
+        heat_flux = compute_Reynolds_heat_flux_fad(primitive_soln_rans, primitive_soln_gradient_rans,primitive_soln_turbulence_model);
+    }
+    else{
+        std::cout << "ERROR in physics/reynolds_averaged_navier_stokes.cpp --> dissipative_flux_templated(): real2!=real || real2!=FadType)" << std::endl;
+        std::abort();
+    }
+    
+    // Step 4: Construct viscous flux; Note: sign corresponds to LHS
+    std::array<dealii::Tensor<1,dim,real2>,dim+2> viscous_flux_rans
+        = this->navier_stokes_physics->dissipative_flux_given_velocities_viscous_stress_tensor_and_heat_flux(vel,viscous_stress_tensor,heat_flux);
+
+    std::array<dealii::Tensor<1,dim,real2>,nstate-(dim+2)> viscous_flux_turbulence_model
+        = this->dissipative_flux_turbulence_model(primitive_soln_rans,primitive_soln_turbulence_model,primitive_soln_gradient_turbulence_model);
+
+    std::array<dealii::Tensor<1,dim,real2>,nstate> viscous_flux;
+    for(int flux_dim=0; flux_dim<dim+2; flux_dim++)
+    {
+        viscous_flux[flux_dim] = viscous_flux_rans[flux_dim];
+    }
+    for(int flux_dim=dim+2; flux_dim<nstate; flux_dim++)
+    {
+        viscous_flux[flux_dim] = viscous_flux_turbulence_model[flux_dim-(dim+2)];
+    }
+    
+    return viscous_flux;
+}
+//To do later
+//Introducing new template to build general extract function for RANS and turbulence model
+//----------------------------------------------------------------
+template <int dim, int nstate, typename real>
+template <typename real2>
+std::array<real2,dim+2> ReynoldsAveragedNavierStokesBase<dim,nstate,real>
+::extract_rans_conservative_solution (
+    const std::array<real2,nstate> &conservative_soln) const
+{   
+    std::array<real2,dim+2> conservative_soln_rans;
+    for(int i=0; i<dim+2; i++){
+        conservative_soln_rans[i] = conservative_soln[i];
+    }
+ 
+    return conservative_soln_rans;
+}
+//To do later
+//Introducing new template to build general extract function for RANS and turbulence model
+//----------------------------------------------------------------
+template <int dim, int nstate, typename real>
+template <typename real2>
+std::array<dealii::Tensor<1,dim,real2>,dim+2> ReynoldsAveragedNavierStokesBase<dim,nstate,real>
+::extract_rans_solution_gradient (
```

Comment:
isn't this function actually extracting the "navier-stokes" solution gradient? (same for the other `extract_rans_...` function)

Reply:
Ah okay, because when I think of a rans solution gradient, I imagine it including the turbulence working variable, so having dim+3 instead of dim+2

Reply:
you can resolve this, maybe just add a comment in the code to clarify this for everyone else

Reply:
fixed.
additional comment is added to clarify